Add new SMS API concatenation tutorial

marklewin opened the pull request:
## Description

Added a new tutorial. The source code is [here](https://github.com/Nexmo/sms-node-concat-tutorial).

## Deploy Notes

Check in sample app.

mheap commented on _tutorials/receiving-concat-sms.md:
```diff
@@ -0,0 +1,398 @@
+
+## In this tutorial
+
+In this tutorial, you create a simple Node.js application using the Express framework that receives inbound SMS via a webhook and determines whether the message is a single-part or multi-part SMS.
```
you *will* create?

marklewin replied:
Agreed!

mheap commented on _tutorials/receiving-concat-sms.md:
```diff
@@ -0,0 +1,398 @@
+
+# Receiving Concatenated SMS
+
+SMS messages that [exceed a certain length](/messaging/sms/guides/concatenation-and-encoding) are split into multiple SMS and transmitted seperately. The receiver must gather all the constituent parts and reassemble them to display the full message.
```
Is it multiple SMS, or is it a single SMS comprised of multiple parts?

marklewin replied:
Reworded

mheap commented on _tutorials/receiving-concat-sms.md:
````diff
@@ -0,0 +1,398 @@
+Download and install [ngrok](https://ngrok.com), then start it with the following command:
+
+```sh
+$ ./ngrok http 5000
````
I think most people install `ngrok` globally, in which case `./ngrok` won't work (but `ngrok` will)

marklewin replied:
Changed to `ngrok`

mheap commented on _tutorials/receiving-concat-sms.md:
```diff
@@ -0,0 +1,398 @@
+
+Now, if any of your virtual numbers receive an SMS, Nexmo will call that webhook endpoint with the message details.
+
+## Send a test SMS
```
Does this work? I didn't think we had Nexmo->Nexmo routing set up.

In the event that it does, I still think that we could cut this section and have the user send an SMS from their personal device to their Nexmo number. This removes the requirement for the Nexmo client lib and any authentication credentials

marklewin replied:
This _does_ work. I agree though that it will make everything simpler if we just get readers to send SMS from their mobile devices.

mheap replied:
We had a chat about this on Slack and the official line is that it's "not supported", so I think we should cut it and get them to send an SMS from their personal device

marklewin replied:
Done!

mheap commented on _tutorials/receiving-concat-sms.md:
```diff
@@ -0,0 +1,398 @@
+
+To enable us to present such messages to our users in the format they were intended, we need to detect if an incoming message has been split in this way and then reassemble it from the parts.
+
+> Notice in the above output that the parts did not arrive in the correct order. This is not uncommon so we need to code our webhook to handle this eventuality.
```
This is a great example 🙌

mheap commented on _tutorials/receiving-concat-sms.md:
```diff
@@ -0,0 +1,398 @@
+function handleInboundSms(request, response) {
+    const params = Object.assign(request.query, request.body);
+
+    if (params['concat'] === 'true') {
```
The strict comparison here could cause issues if the value becomes a boolean rather than a string (similar things have happened in the past). `if (params['concat'])` should be enough

marklewin replied:
Noted, and changed.

mheap replied:
`if (params['concat'])` works for truthy values like strings too

```
> x = "true"
> if (x) { console.log("Foo"); }
Foo

> x = false
false
> if (x) { console.log("Foo"); }
undefined
```

mheap commented on _tutorials/receiving-concat-sms.md:
```diff
@@ -0,0 +1,398 @@
+
+Before we even attempt to reassemble the message from its parts, we need to ensure that we have all the parts for a given message reference. Remember that there is no guarantee that all the parts will arrive in the correct order, so it is not simply a matter of checking if `concart-part` equals `concat-total`.
+
+We can do this by filtering the `concat_sms` array to include only those SMS objects that share the same `concat-ref` as the SMS that we have just received. If the length of that filtered array is the same as `concat-total`, then we have all the parts for that message and can then reassemble them:
```
Good catch! I'd never have thought about the situation where multiple concatenated messages arrive in the same time frame

marklewin replied:
:)

marklewin commented:
Incorporated changes requested by @mheap and updated code samples (and accompanying tutorial code [here](https://github.com/Nexmo/sms-node-concat-tutorial)) to ES6.
